security/pkg/pki/ca/ca.go

```diff
@@ -160,8 +160,16 @@ func NewSelfSignedIstioCAOptions(ctx context.Context,
 		// Write the key/cert back to secret so they will be persistent when CA restarts.
 		secret := k8ssecret.BuildSecret("", CASecret, namespace, nil, nil, nil, pemCert, pemKey, istioCASecretType)
 		if _, err = client.Secrets(namespace).Create(context.TODO(), secret, metav1.CreateOptions{}); err != nil {
-			pkiCaLog.Errorf("Failed to write secret to CA (error: %s). Abort.", err)
-			return nil, fmt.Errorf("failed to create CA due to secret write error")
+			// From https://github.com/istio/istio/issues/28236, in rare occasions,
+			// creating secret may fail due to: (error: secrets "istio-ca-secret" already exists).
+			// Based on https://github.com/istio/istio/issues/28236, when creation failed, try read the secret,
+			// if the reading also fails, return the error then.
+			_, readErr := client.Secrets(namespace).Get(context.TODO(), CASecret, metav1.GetOptions{})
```

Comment:
If we get to this state, and `istio-ca-secret` in cluster isn't consistent with our local (in memory) self generated cert, I think we may end up in an inconsistent state and things will break?

Reply:
Do we have a design document for Multi Control Plane? Based on https://github.com/istio/istio/issues/28236, it seems multiple control planes are writing to the same secret. Does the design of Multi Control Plane support such workflows?

Reply:
@lei-tang the issue reported is for multiple control plane (design doc: https://docs.google.com/document/d/1d1z5PC8wvh9QiR3NesAaPm6w2HnE4-KIfjB7GB-_Wf8/edit#), but the fundamental issue would happen with 2 instances of istiod in the same deployment as well. There is a clear race condition on startup between any 2 instances.

Reply:
Thanks for the link to the design document! As multiple control planes will generate multiple CA certs (each for one control plane), multiple local CA certs will likely create TLS problems. Thus, the right solution is to design and implement a mechanism that multiple control planes share a single CA certificate rather than having multiple CA certificate.

Reply:
> Thanks for the link to the design document! As multiple control planes will generate multiple CA certs (each for one control plane), multiple local CA certs will likely create TLS problems. Thus, the right solution is to design and implement a mechanism that multiple control planes share a single CA certificate rather than having multiple CA certificates.


The control planes DO all share a single CA certificate. Please note we shipped that design in Istio 1.6. This issue is not even strictly related to it... if you have any two istiod instances startup at the same time (whether they are in the same revision or not), this bug can occur. Its a simple race condition:

* At startup, if the ca is missing, istiod will create it
* If two instances detect there is no ca, they will both attempt to create it, which fails

Reply:
I revised my comment above to remove the CA part: what I mean is multiple control planes will generate multiple certs (each for one control plane); I do not mean they do not share the same CA cert. As multiple control planes use their own certs, the TLS clients may not recognize the certs of multiple control planes. To enable TLS clients recognize the certs of multiple control planes, we may need to design and implement a mechanism to enable multiple control planes to share the same cert.

Reply:
> As multiple control planes use their own certs

They use the same cert

Reply:
They are not independent meshes, they are all part of the same mesh/trust. Just two different instances of istiod.

Reply:
> > As multiple control planes use their own certs
> 
> They use the same cert

If multiple control planes use the same cert, the fix in this PR should work. But IIUC, multiple control planes generate their own certs, then use their own certs for hosting TLS connections.

Reply:
I don't think it will?

Today:
1) 2 Istiod startup, they try to read the ca secret: https://github.com/istio/istio/blob/32a5c49a89e0360bcc2a0fae23bff0cbd53b3f34/security/pkg/pki/ca/ca.go#L100. This fails since there is none yet
2) They both generate a new root cert: https://github.com/istio/istio/blob/32a5c49a89e0360bcc2a0fae23bff0cbd53b3f34/security/pkg/pki/ca/ca.go#L146
3) They both try to write it, ones fail and exits: https://github.com/istio/istio/blob/32a5c49a89e0360bcc2a0fae23bff0cbd53b3f34/security/pkg/pki/ca/ca.go#L162. This is what the issue is about 

With this PR:
1) 2 Istiod startup, they try to read the ca secret: https://github.com/istio/istio/blob/32a5c49a89e0360bcc2a0fae23bff0cbd53b3f34/security/pkg/pki/ca/ca.go#L100. This fails since there is none yet
2) They both generate a new root cert: https://github.com/istio/istio/blob/32a5c49a89e0360bcc2a0fae23bff0cbd53b3f34/security/pkg/pki/ca/ca.go#L146
3) They both try to write it, and one fails but silently continues.
4) The 2 instances now have a different root cert

btw, this also means if we have temporary outage of k8s, the read can fail and the write succeeds. In this case, we would also have a different root cert in two instances

Reply:
I revised the PR as follows: upon failing to create the secret, an istiod instance reads the cert from the secret, and uses it as its cert if the secret exists.  With the revision, the flow for failing to create the secret becomes:

1. Two Istiod startup, they try to read the ca secret. This fails since there is none yet
2. They both generate a new cert.
3. They both try to write their certs to the secret, and one fails due to (error: secrets "istio-ca-secret" already exists) but continues to read the cert from the secret and uses it as its.
4.  The two instances now use the same cert.